File: xju/Time.cc
//     -*- mode: c++ ; c-file-style: "osse" ; -*-
//
// Copyright (c) 1997-2007 Trevor Taylor
//
// Permission to use, copy, modify, distribute and sell this software
// and its documentation for any purpose is hereby granted without fee,
// provided that the above copyright notice appear in all.
// Trevor Taylor makes no representations about the suitability of this
// software for any purpose.  It is provided "as is" without express or
// implied warranty.
//
#include <xju/Time.hh>


#include <sys/time.h>
#include <iomanip>
#include <iostream>
#include <stdlib.h>
#include <xju/assert.hh>
#include <functional>
#include <sys/types.h>


namespace xju
{
    Time::Time(unsigned long secs, unsigned int usecs) throw():
        _secs(secs),
        _usecs(usecs)
    {
        xju::assert_(usecs, std::less<unsigned int>(), 1000000U);
    }
    
    bool Time::operator==(const Time &b) const throw()
    {
        return (_secs == b._secs && _usecs == b._usecs);
    }
    
    bool Time::operator<(const Time &b) const throw()
    {
        if (_secs > b._secs)
            return false;
        
        if (_secs < b._secs)
            return true;
        
        if (_usecs < b._usecs)
            return true;
        
        return false;
    }
    bool Time::operator>(const Time& x) const throw()
    {
        return !operator<(x) && !operator==(x); 
    }
    bool Time::operator<=(const Time& x) const throw()
    {
        return operator<(x) || operator==(x);
    }
    bool Time::operator>=(const Time& x) const throw()
    {
        return !operator<(x);
    }
    
    
    unsigned long Time::getSecs() const throw()
    {
        return _secs;
    }
    
    unsigned int Time::getUsecs() const throw()
    {
        return _usecs;
    }
    
    Time Time::now() throw()
    {
#ifndef __MINGW32__
        timeval now;
        gettimeofday(&now, 0);
        
        //
        // timeval.tv_sec is typically signed, but we're
        // converting it to unsigned
        //
        xju::assert_(now.tv_sec, std::greater_equal<time_t>(), 0);
        
        return Time(now.tv_sec, now.tv_usec);
#else
        time_t now;
        time(&now);
// REVISIT get usec from somewhere
        return Time(now, 0);
#endif
    }
    
}
xju::Time operator+(const xju::Time &t1, const xju::MicroSeconds &t2) throw()
{
    unsigned int usecs = t1.getUsecs() + t2.value()%1000000;
    unsigned long secs = t1.getSecs() + t2.value()/1000000 + (usecs / 1000000U);
    usecs = usecs % 1000000U;
    return xju::Time(secs, usecs);
}

xju::Time operator-(xju::Time const& t, 
                    xju::MicroSeconds const& d) throw()
{
    xju::MicroSeconds const a(t - xju::Time(0, 0));
    xju::MicroSeconds const b(a - d);
    return xju::Time(b.value() / 1000000, b.value() % 1000000);
}

xju::MicroSeconds operator-(const xju::Time &t1, const xju::Time &t2) throw()
{
    xju::assert_(t1, std::greater_equal<xju::Time>(), t2);
    unsigned long secs = 0;
    unsigned int usecs = 0;
    
    if (t1.getUsecs() >= t2.getUsecs()) {
        usecs = t1.getUsecs() - t2.getUsecs();
        secs = t1.getSecs() - t2.getSecs();
    } else { // t1.getUsecs() < t2.getUsecs()
        usecs = (1000000U - t2.getUsecs()) + t1.getUsecs();
        secs = (t1.getSecs()-t2.getSecs()-1UL);
    }
    return xju::MicroSeconds(((uint64_t)secs)*1000000+usecs);
}

std::ostream& operator<<(std::ostream& s, const xju::Time& x) throw()
{
    s << x.getSecs() << "." << xju::format::int_(x.getUsecs(), 6);
    return s;
}

File: xju/time.py
# Copyright (c) 2022 Trevor Taylor
# coding: utf-8
# 
# Permission to use, copy, modify, and/or distribute this software for
# any purpose with or without fee is hereby granted, provided that all
# copyright notices and this permission notice appear in all copies.
# 
# THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
# WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
# MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
# ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
# WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
# ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
# OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
#
from xju.newtype import Int,Float,Union
from typing import overload, Self, Final
import time
import asyncio
from dataclasses import dataclass
from typing import Sequence

class HoursTag:pass
class MinutesTag:pass
class SecondsTag:pass

class Hours(Int[HoursTag]): pass
class Minutes(Int[MinutesTag]): pass
class Seconds(Int[SecondsTag]): pass

class DurationTag:pass
class Duration(Float[DurationTag]):pass


@dataclass(unsafe_hash=True,order=True)
class Timestamp():
    __value:Final[float]
    def __init__(self, value:float|int):
        self.__value=float(value)
    def value(self):
        return self.__value
    def __str__(self):
        return str(self.__value)
    def __repr__(self):
        return repr(self.__value)
    def __format__(self,f:str)->str:
        return self.__value.__format__(f)
    def __add__(self, x:Union[Duration,Hours,Minutes,Seconds]):
        if isinstance(x,Hours):
            return Timestamp(self.value()+x.value()*3600)
        if isinstance(x,Minutes):
            return Timestamp(self.value()+x.value()*60)
        if isinstance(x,Seconds):
            return Timestamp(self.value()+x.value())
        return Timestamp(self.value()+x.value())
    def __radd__(self, duration:Union[Duration,Hours,Minutes,Seconds]):
        return self.__add__(duration)

    @overload
    def __sub__(self, t2:Self)->Duration:
        pass
    @overload
    def __sub__(self, duration:Union[Duration,Hours,Minutes,Seconds])->Self:
        pass
    def __sub__(self, x):
        if isinstance(x,Duration):
            return Timestamp(self.value()-x.value())
        if isinstance(x,Hours):
            return Timestamp(self.value()-x.value()*3600)
        if isinstance(x,Minutes):
            return Timestamp(self.value()-x.value()*60)
        if isinstance(x,Seconds):
            return Timestamp(self.value()-x.value())
        if isinstance(x,Timestamp):
            return Duration(self.value()-x.value())
        x_type=type(x)
        assert False, f'cannot subtract {x} of type {x_type} from Timestamp'
    def __float__(self):
        return self.__value
    pass

def now()->Timestamp:
    return Timestamp(time.time())

def sleep_for(x:Duration):
    time.sleep(x.value())
    pass

async def async_sleep_until(deadline: Timestamp) -> None:
    """sleep while we're not yet passed {deadline}"""
    await asyncio.sleep(max(0.0, (deadline-now()).value()))

async def async_sleep_until_first_of(a_deadline: Timestamp,
                                     *more_deadlines: Timestamp) -> None:
    deadlines:list[Timestamp]=[a_deadline]+list(more_deadlines)
    deadlines.sort()
    await async_sleep_until(deadlines.pop(0))
